### PR TITLE
violation of duplicate when cascading

### PR DESCRIPTION
**Pull request by wy193777**
Related to #2116.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-version: '3'
+version: '2'
```
Comment (pleerock) [inline docker-compose.yml]: remove this from commit please

File: test/github-issues/2116/issue-2116.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        const photo1 = new Photo();
+        photo1.description = "Tall trees";
+        photo1.user = user1;
+
+        const photo2 = new Photo();
+        photo2.description = "Short trees";
+        photo2.user = user2;
+        try {
+            await connection.transaction(async tManager => {
+                tManager.save(photo1);
```
Comment (pleerock) [inline test/github-issues/2116/issue-2116.ts]: you missed await before save

File: test/github-issues/2116/issue-2116.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            });
+
+            await connection.transaction(async tManager => {
+                tManager.save(photo2);
```
Comment (pleerock) [inline test/github-issues/2116/issue-2116.ts]: you missed await before save

**Comment by wy193777** [comment]
Emmmm, still cannot reproduce the same error...

**Comment by wy193777** [comment]
My test doesn't work on Postgres, but passed when using mssql. Maybe this is a bug.

But still not the bug I'm trying to reproduce. Let me try a more complex example.

Should I keep this test? Because seems I find an inconsistency between Postgres and MSSQL.

**Comment by pleerock** [comment]
mssql is being run on CI, so if you test using CI then return back to postgres or better to remove enabled drivers at all

**Comment by wy193777** [comment]
I reproduced the error locally with this code. I guess a database contains the data we want to insert is needed to get the error I mentioned. Not sure how to achieve this on on the test case on Travis.

I post the error message here：
```

> typeorm@0.2.5 test /home/shenghan/code/typeorm
> tsc && mocha --file ./build/compiled/test/utils/test-setup.js --bail --recursive --timeout 30000  ./build/compiled/test "--grep=github issues > #2116"



  github issues > #2116 cascade save with duplicates
tedious deprecated The `IntN` data type is internal and will be removed. node_modules/mssql/lib/tedious.js:63:20
tedious deprecated The `BitN` data type is internal and will be removed. node_modules/mssql/lib/tedious.js:71:40
{ QueryFailedError: Error: Violation of UNIQUE KEY constraint 'UQ_2c5aa339240c0c3ae97fcc9dc4c'. Cannot insert duplicate key in object 'dbo.country'. The duplicate key value is (US).
    at new QueryFailedError (/home/shenghan/code/typeorm/src/error/QueryFailedError.ts:7:9)
    at /home/shenghan/code/typeorm/src/driver/sqlserver/SqlServerQueryRunner.ts:209:37
    at _query (/home/shenghan/code/typeorm/node_modules/mssql/lib/base.js:1253:25)
    at Request.tds.Request.err [as userCallback] (/home/shenghan/code/typeorm/node_modules/mssql/lib/tedious.js:620:15)
    at Request._this.callback (/home/shenghan/code/typeorm/node_modules/tedious/lib/request.js:60:27)
    at Connection.endOfMessageMarkerReceived (/home/shenghan/code/typeorm/node_modules/tedious/lib/connection.js:1777:20)
    at Connection.dispatchEvent (/home/shenghan/code/typeorm/node_modules/tedious/lib/connection.js:863:38)
    at Parser.<anonymous> (/home/shenghan/code/typeorm/node_modules/tedious/lib/connection.js:684:18)
    at Parser.emit (events.js:182:13)
    at Parser.<anonymous> (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/token-stream-parser.js:54:15)
    at Parser.emit (events.js:182:13)
    at addChunk (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_readable.js:291:12)
    at readableAddChunk (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_readable.js:278:11)
    at Parser.Readable.push (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_readable.js:245:10)
    at Parser.Transform.push (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_transform.js:148:32)
    at Parser.afterTransform (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_transform.js:91:10)
    at Parser._transform (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:69:9)
    at Parser.Transform._read (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_transform.js:184:10)
    at Parser.Transform._write (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_transform.js:172:83)
    at doWrite (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_writable.js:418:64)
    at writeOrBuffer (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_writable.js:407:5)
    at Parser.Writable.write (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_writable.js:334:11)
    at Parser.addEndOfMessageMarker (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/token-stream-parser.js:80:26)
    at Connection.message (/home/shenghan/code/typeorm/node_modules/tedious/lib/connection.js:1768:32)
    at Connection.dispatchEvent (/home/shenghan/code/typeorm/node_modules/tedious/lib/connection.js:863:38)
    at MessageIO.<anonymous> (/home/shenghan/code/typeorm/node_modules/tedious/lib/connection.js:757:18)
    at MessageIO.emit (events.js:182:13)
    at ReadablePacketStream.<anonymous> (/home/shenghan/code/typeorm/node_modules/tedious/lib/message-io.js:102:16)
    at ReadablePacketStream.emit (events.js:182:13)
    at addChunk (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_readable.js:291:12)
    at readableAddChunk (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_readable.js:278:11)
    at ReadablePacketStream.Readable.push (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_readable.js:245:10)
    at ReadablePacketStream.Transform.push (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_transform.js:148:32)
    at ReadablePacketStream._transform (/home/shenghan/code/typeorm/node_modules/tedious/lib/message-io.js:70:16)
    at ReadablePacketStream.Transform._read (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_transform.js:184:10)
    at ReadablePacketStream.Transform._write (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_transform.js:172:83)
    at doWrite (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_writable.js:418:64)
    at writeOrBuffer (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_writable.js:407:5)
    at ReadablePacketStream.Writable.write (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_writable.js:334:11)
    at Socket.ondata (_stream_readable.js:672:20)
    at Socket.emit (events.js:182:13)
    at addChunk (_stream_readable.js:280:12)
    at readableAddChunk (_stream_readable.js:265:11)
    at Socket.Readable.push (_stream_readable.js:220:10)
    at TCP.onread (net.js:638:20)
  message: 'Error: Violation of UNIQUE KEY constraint \'UQ_2c5aa339240c0c3ae97fcc9dc4c\'. Cannot insert duplicate key in object \'dbo.country\'. The duplicate key value is (US).',
  code: 'EREQUEST',
  number: 2627,
  lineNumber: 1,
  state: 1,
  class: 14,
  serverName: '90e3ec93ea98',
  procName: '',
  originalError: 
   { Error: Violation of UNIQUE KEY constraint 'UQ_2c5aa339240c0c3ae97fcc9dc4c'. Cannot insert duplicate key in object 'dbo.country'. The duplicate key value is (US).
    at handleError (/home/shenghan/code/typeorm/node_modules/mssql/lib/tedious.js:517:19)
    at Connection.emit (events.js:182:13)
    at Parser.<anonymous> (/home/shenghan/code/typeorm/node_modules/tedious/lib/connection.js:490:16)
    at Parser.emit (events.js:182:13)
    at Parser.<anonymous> (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/token-stream-parser.js:54:15)
    at Parser.emit (events.js:182:13)
    at addChunk (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_readable.js:291:12)
    at readableAddChunk (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_readable.js:278:11)
    at Parser.Readable.push (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_readable.js:245:10)
    at Parser.Transform.push (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_transform.js:148:32)
    at doneParsing (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:110:18)
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/infoerror-token-parser.js:46:5
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/infoerror-token-parser.js:13:19
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:247:9
    at Parser.awaitData (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:144:9)
    at Parser.readUInt32LE (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:244:12)
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/infoerror-token-parser.js:12:90
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:446:11
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:433:9
    at Parser.awaitData (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:144:9)
    at Parser.readBuffer (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:430:12)
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:445:17
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:170:9
    at Parser.awaitData (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:144:9)
    at Parser.readUInt8 (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:167:12)
    at Parser.readBVarChar (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:444:12)
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/infoerror-token-parser.js:11:22
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:446:11
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:433:9
    at Parser.awaitData (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:144:9)
    at Parser.readBuffer (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:430:12)
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:445:17
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:170:9
    at Parser.awaitData (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:144:9)
    at Parser.readUInt8 (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:167:12)
    at Parser.readBVarChar (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:444:12)
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/infoerror-token-parser.js:10:20
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:460:11
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:433:9
    at Parser.awaitData (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:144:9)
    at Parser.readBuffer (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:430:12)
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:459:17
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:203:9
    at Parser.awaitData (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:144:9)
    at Parser.readUInt16LE (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:200:12)
    at Parser.readUsVarChar (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:458:12)
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/infoerror-token-parser.js:9:18
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:170:9
    at Parser.awaitData (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:144:9)
    at Parser.readUInt8 (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:167:12)
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/infoerror-token-parser.js:8:16
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:170:9
    at Parser.awaitData (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:144:9)
    at Parser.readUInt8 (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:167:12)
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/infoerror-token-parser.js:7:14
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:247:9
    at Parser.awaitData (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:144:9)
    at Parser.readUInt32LE (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:244:12)
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/infoerror-token-parser.js:6:12
    at /home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:203:9
    at Parser.awaitData (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:144:9)
    at Parser.readUInt16LE (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:200:12)
    at parseToken (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/infoerror-token-parser.js:5:10)
    at Object.errorParser (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/infoerror-token-parser.js:43:3)
    at Parser.parseTokens (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:120:29)
    at Parser._transform (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/stream-parser.js:95:14)
    at Parser.Transform._read (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_transform.js:184:10)
    at Parser.Transform._write (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_transform.js:172:83)
    at doWrite (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_writable.js:418:64)
    at writeOrBuffer (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_writable.js:407:5)
    at Parser.Writable.write (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_writable.js:334:11)
    at Parser.addBuffer (/home/shenghan/code/typeorm/node_modules/tedious/lib/token/token-stream-parser.js:69:26)
    at Connection.sendDataToTokenStreamParser (/home/shenghan/code/typeorm/node_modules/tedious/lib/connection.js:1013:37)
    at Connection.data (/home/shenghan/code/typeorm/node_modules/tedious/lib/connection.js:1757:24)
    at Connection.dispatchEvent (/home/shenghan/code/typeorm/node_modules/tedious/lib/connection.js:863:38)
    at MessageIO.<anonymous> (/home/shenghan/code/typeorm/node_modules/tedious/lib/connection.js:754:18)
    at MessageIO.emit (events.js:182:13)
    at ReadablePacketStream.<anonymous> (/home/shenghan/code/typeorm/node_modules/tedious/lib/message-io.js:100:14)
    at ReadablePacketStream.emit (events.js:182:13)
    at addChunk (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_readable.js:291:12)
    at readableAddChunk (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_readable.js:278:11)
    at ReadablePacketStream.Readable.push (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_readable.js:245:10)
    at ReadablePacketStream.Transform.push (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_transform.js:148:32)
    at ReadablePacketStream._transform (/home/shenghan/code/typeorm/node_modules/tedious/lib/message-io.js:70:16)
    at ReadablePacketStream.Transform._read (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_transform.js:184:10)
    at ReadablePacketStream.Transform._write (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_transform.js:172:83)
    at doWrite (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_writable.js:418:64)
    at writeOrBuffer (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_writable.js:407:5)
    at ReadablePacketStream.Writable.write (/home/shenghan/code/typeorm/node_modules/tedious/node_modules/readable-stream/lib/_stream_writable.js:334:11)
    at Socket.ondata (_stream_readable.js:672:20)
    at Socket.emit (events.js:182:13)
    at addChunk (_stream_readable.js:280:12)
    at readableAddChunk (_stream_readable.js:265:11)
    at Socket.Readable.push (_stream_readable.js:220:10)
    at TCP.onread (net.js:638:20)
     info: 
      { number: 2627,
        state: 1,
        class: 14,
        message: 'Violation of UNIQUE KEY constraint \'UQ_2c5aa339240c0c3ae97fcc9dc4c\'. Cannot insert duplicate key in object \'dbo.country\'. The duplicate key value is (US).',
        serverName: '90e3ec93ea98',
        procName: '',
        lineNumber: 1,
        name: 'ERROR',
        event: 'errorMessage' } },
  name: 'QueryFailedError',
  precedingErrors: [],
  query: 'INSERT INTO "country"("name") OUTPUT INSERTED."id" VALUES (@0)',
  parameters: 
   [ MssqlParameter { value: 'US', type: 'nvarchar', params: [] } ] }
    1) Save query should fail on second


  0 passing (319ms)
  1 failing

  1) github issues > #2116 cascade save with duplicates
       Save query should fail on second:
     TypeError: Cannot read property 'state' of undefined
      at Object.<anonymous> (test/github-issues/2116/issue-2116.ts:60:19)
      at step (build/compiled/test/github-issues/2116/issue-2116.js:32:23)
      at Object.next (build/compiled/test/github-issues/2116/issue-2116.js:13:53)
      at fulfilled (build/compiled/test/github-issues/2116/issue-2116.js:4:58)
      at process._tickCallback (internal/process/next_tick.js:178:7)



npm ERR! Test failed.  See above for more details.

```

**Comment by wy193777** [comment]
Any document about how to debug a test with IDE like VSCode? Logging isn't enough for me to identify which part of the code related to the problem. The strategy on how the framework load data from the database before starting saving might be the problem.

After looking into the code I guess that one manager.save() with cascading might cause several requests to the database, so simply check if the data already in the database before cascading save will not create too much overhead.

**Pull request closed without merging**